charmpath.go

```diff
@@ -0,0 +1,46 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmrepo
+
+import (
+	"os"
+	"path/filepath"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+)
+
+// NewCharmAtPath returns the charm represented by this path.
+// If the series is not specified, the charm's default
+// series is used, if any. Otherwise, the series is
+// validated against those the charm declares it supports.
+func NewCharmAtPath(path, series string) (charm.Charm, *charm.URL, error) {
+	if path == "" {
+		return nil, nil, errgo.New("path to charm not specified")
+	}
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return nil, nil, os.ErrNotExist
+	}
+	ch, err := charm.ReadCharm(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil, CharmNotFound(path)
+		}
+		return nil, nil, err
+	}
+	_, name := filepath.Split(path)
+	meta := ch.Meta()
+	seriesToUse, err := charm.SeriesForCharm(series, meta.Series)
+	if err != nil {
+		return nil, nil, err
+	}
+	url := &charm.URL{
+		Schema:   "local",
+		Name:     name,
+		Series:   seriesToUse,
```

Comment:
I think this can return a URL with an empty series, as SeriesForCharm can return "", nil when meta.Series and series are empty. That breaks the invariant that charm.URL is supposed to maintain. I'd suggest returning *charm.Reference from this method.

As I've mentioned before we should seriously consider getting rid of charm.Reference entirely and relaxing the series-not-empty constraint on charm.URL.